source/print/ur_print.cpp

```diff
@@ -0,0 +1,3875 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file ur_print.cpp
+ *
+ */
+
+#include "ur_print.h"
+#include "ur_print.hpp"
+
+#include <sstream>
+#include <string.h>
+
+using namespace ur_print;
+
+void strncpy(std::stringstream *ss, char *out, const size_t size) {
+#if defined(_WIN32)
+    strncpy_s(out, size, ss->str().c_str(), ss->str().size() + 1);
+#else
+    strncpy(out, ss->str().c_str(), size);
+#endif
+}
+
+ur_result_t urPrintFunction(enum ur_function_t value, char *out,
```

Comment:
there's no way for the user to know how big the buffer should be. A common solution is to either add a pointer where the function will write the length of the output or make the `size` parameter pull double duty as input when out is provided and output when it's null.